Modules/gcmodule.c

```diff
@@ -1191,80 +1251,260 @@ handle_resurrected_objects(PyGC_Head *unreachable, PyGC_Head* still_unreachable,
+    return gc_old_space(gc) != aging_space;
+}
+
+struct container_and_flag {
+    PyGC_Head *container;
+    uintptr_t aging_space;
+};
 
-    if (gcstate->debug & DEBUG_STATS) {
-        PySys_WriteStderr("gc: collecting generation %d...\n", generation);
-        show_stats_each_generations(gcstate);
-        t1 = _PyTime_GetPerfCounter();
+/* A traversal callback for adding to container) */
+static int
+visit_add_to_container(PyObject *op, void *arg)
+{
+    OBJECT_STAT_INC(object_visits);
+    struct container_and_flag *cf = (struct container_and_flag *)arg;
+    uintptr_t aging = cf->aging_space;
+    if (_PyObject_IS_GC(op)) {
+        PyGC_Head *gc = AS_GC(op);
+        if (_PyObject_GC_IS_TRACKED(op) &&
+            gc_old_space(gc) != aging) {
+            assert(!_Py_IsImmortal(op));
+            gc_flip_old_space(gc);
+            gc_list_move(gc, cf->container);
+        }
     }
+    return 0;
+}
 
-    if (PyDTrace_GC_START_ENABLED())
-        PyDTrace_GC_START(generation);
+static uintptr_t
+expand_region_transitively_reachable(PyGC_Head *container, GCState *gcstate)
+{
+    PyGC_Head *next;
+    PyGC_Head *gc = GC_NEXT(container);
+    validate_list(container, collecting_clear_unreachable_clear);
 
-    /* update collection and allocation counters */
-    if (generation+1 < NUM_GENERATIONS)
-        gcstate->generations[generation+1].count += 1;
-    for (i = 0; i <= generation; i++)
-        gcstate->generations[i].count = 0;
+    struct container_and_flag arg = {
+        .container = container,
+        .aging_space = gcstate->aging_space,
+    };
+    uintptr_t size = 0;
+    while (gc != container) {
+        assert(!IS_IN_OLDEST(gc, gcstate->aging_space));
+        if (_Py_IsImmortal(FROM_GC(gc))) {
+            next = GC_NEXT(gc);
+            gc_list_move(gc, &get_gc_state()->permanent_generation.head);
+            gc = next;
+            continue;
+        }
+        PyObject *op = FROM_GC(gc);
+        traverseproc traverse = Py_TYPE(op)->tp_traverse;
+        (void) traverse(op,
+                        visit_add_to_container,
+                        &arg);
+        gc = GC_NEXT(gc);
+        size++;
+    }
+    return size;
+}
 
-    /* merge younger generations with one we are currently collecting */
-    for (i = 0; i < generation; i++) {
-        gc_list_merge(GEN_HEAD(gcstate, i), GEN_HEAD(gcstate, generation));
+static void
+gc_collect_increment(PyThreadState *tstate, struct gc_collection_stats *stats)
+{
+    GCState *gcstate = &tstate->interp->gc;
+    /* We only have work to do if incremental_gc_progress is negative */
+    if (gcstate->incremental_gc_progress >= 0) {
+        return;
     }
+    PyGC_Head *oldest = &gcstate->old[gcstate->aging_space^1].head;
+    PyGC_Head *aging = &gcstate->old[gcstate->aging_space].head;
+    PyGC_Head increment;
+    gc_list_init(&increment);
+    /* Multiply by 4 so that the default incremental threshold of 10
```

Comment:
Maybe put this in a const?  There's a 2nd magic `4` down below which I assume is tied to this number but not documented or linked in any way.